gnome-initial-setup/pages/site/gis-site-page.c

```diff
@@ -0,0 +1,473 @@
+  const gchar *region = gtk_entry_get_text (GTK_ENTRY (priv->region_entry));
+  const gchar *country = gtk_entry_get_text (GTK_ENTRY (priv->country_entry));
+
+  return gis_site_new (id, facility, street, locality, region, country);
+}
+
+static void
+ini_string_append_safe (GString     *string,
+                        const gchar *property,
+                        const gchar *value)
+{
+  g_string_append_printf (string, "%s = %s\n", property, value ? value : "");
+}
+
+static gchar*
+ini_file_content_from_site (GisSite *site)
+{
+  GString *builder = g_string_new ("[Label]\n");
+
+  // Note the field names aren't 1:1 between GisSite and the file format
+  ini_string_append_safe (builder, "id", gis_site_get_id (site));
+  ini_string_append_safe (builder, "facility", gis_site_get_facility (site));
+  ini_string_append_safe (builder, "street", gis_site_get_street (site));
+  ini_string_append_safe (builder, "city", gis_site_get_locality (site));
+  ini_string_append_safe (builder, "state", gis_site_get_region (site));
+  ini_string_append_safe (builder, "country", gis_site_get_country (site));
+
+  return g_string_free (builder, FALSE);
+}
+
+static void
+gis_site_page_save_data (GisPage *gis_page)
+{
+  GisSitePage *page = GIS_SITE_PAGE (gis_page);
+  gchar *location_file_content;
+  gchar *filename_created;
+  GError *error_open = NULL;
+  GError *error_pkexec = NULL;
+  GisSite *site;
+  int fd;
+
+  if (gis_page->driver == NULL)
+    return;
+
+  site = get_site_from_widgets (page);
+  location_file_content = ini_file_content_from_site (site);
+
+  fd = g_file_open_tmp (LOCATION_FILENAME_TEMPL, &filename_created,
+                        &error_open);
+  if (error_open == NULL)
+    {
+      FILE *temp_file = fdopen (fd, "w");
+
+      if (fputs (location_file_content, temp_file) >= 0)
+        {
+          if (!fclose (temp_file))
+            {
+              if (!gis_pkexec (LIBEXECDIR "/eos-write-location",
+                               filename_created, NULL, &error_pkexec))
+                {
+                  g_warning ("%s", error_pkexec->message);
+                  g_error_free (error_pkexec);
+                }
+            }
+        }
+      else
+        {
+          int saved_errno = errno;
+          g_warning ("Failed to save temporary location file: %s: %s",
+                     filename_created, g_strerror (saved_errno));
+        }
+    }
+  else
+    {
+      g_warning ("Failed to record site information for metrics: %s",
+                 error_open->message);
+      g_error_free (error_open);
+    }
+
+  g_free (filename_created);
+  g_free (location_file_content);
+  g_clear_object (&site);
+}
+
+static void
+entry_site_changed (GObject *object, GParamSpec *param, GisSitePage *page)
+{
+  GisSitePagePrivate *priv = gis_site_page_get_instance_private (page);
+  GisSiteSearchEntry *entry = GIS_SITE_SEARCH_ENTRY (object);
+  GisSite *site;
+
+  site = gis_site_search_entry_get_site (entry);
+  set_site (page, site);
```

Comment:
`gis_site_search_entry_get_site()` returns a new reference that's leaked here in this function. However I don't think that such function needs to return a new reference.